src/main/java/org/opensearch/ad/ratelimit/ConcurrentQueue.java

```diff
@@ -0,0 +1,162 @@
+
+    private Instant lastExecuteTime;
+    private Duration executionTtl;
+
+    /**
+     *
+     * Constructor with dependencies and configuration.
+     *
+     * @param queueName queue's name
+     * @param heapSizeInBytes ES heap size
+     * @param singleRequestSizeInBytes single request's size in bytes
+     * @param maxHeapPercentForQueueSetting max heap size used for the queue. Used for
+     *   rate AD's usage on ES threadpools.
+     * @param clusterService Cluster service accessor
+     * @param random Random number generator
+     * @param adCircuitBreakerService AD Circuit breaker service
+     * @param threadPool threadpool accessor
+     * @param settings Cluster settings getter
+     * @param maxQueuedTaskRatio maximum queued tasks ratio in ES threadpools
+     * @param clock Clock to get current time
+     * @param mediumSegmentPruneRatio the percent of medium priority requests to prune when the queue is full
+     * @param lowSegmentPruneRatio the percent of low priority requests to prune when the queue is full
+     * @param maintenanceFreqConstant a constant help define the frequency of maintenance.  We cannot do
+     *   the expensive maintenance too often.
+     * @param clientUtil utility with ES client
+     * @param concurrencySetting Max concurrent processing of the queued events
+     * @param executionTtl Max execution time of a single request
+     * @param stateTtl max idle state duration.  Used to clean unused states.
+     * @param nodeStateManager node state accessor
+     */
+    public ConcurrentQueue(
+        String queueName,
+        long heapSizeInBytes,
+        int singleRequestSizeInBytes,
+        Setting<Float> maxHeapPercentForQueueSetting,
+        ClusterService clusterService,
+        Random random,
+        ADCircuitBreakerService adCircuitBreakerService,
+        ThreadPool threadPool,
+        Settings settings,
+        float maxQueuedTaskRatio,
+        Clock clock,
+        float mediumSegmentPruneRatio,
+        float lowSegmentPruneRatio,
+        int maintenanceFreqConstant,
+        ClientUtil clientUtil,
+        Setting<Integer> concurrencySetting,
+        Duration executionTtl,
+        Duration stateTtl,
+        NodeStateManager nodeStateManager
+    ) {
+        super(
+            queueName,
+            heapSizeInBytes,
+            singleRequestSizeInBytes,
+            maxHeapPercentForQueueSetting,
+            clusterService,
+            random,
+            adCircuitBreakerService,
+            threadPool,
+            settings,
+            maxQueuedTaskRatio,
+            clock,
+            mediumSegmentPruneRatio,
+            lowSegmentPruneRatio,
+            maintenanceFreqConstant,
+            stateTtl,
+            nodeStateManager
+        );
+
+        this.permits = new Semaphore(concurrencySetting.get(settings));
+        clusterService.getClusterSettings().addSettingsUpdateConsumer(concurrencySetting, it -> permits = new Semaphore(it));
```

Comment:
Is there a possibility where the same concurrencySetting is passed in for 2 separate queue? If so, is this call safe?

Reply:
Our queues are singletons. So it is safe here.  It seems everybody did it this way.  Not sure how ES handles the race conditions of setting change.  There is a complicated protocol to ensure consistency and roll back if not from my read of the related code briefly.

Reply:
I see. So the concurrency setting will be unique for each queue?

Reply:
yes